ARCs/arc-0300.md

```diff
@@ -0,0 +1,71 @@
+author: Bruno Martins (@ehanoc, bruno.martins@algorand.foundation)
+status: Draft
+type: Standards Track
+created: 2024-04-15
+---
+
+# ARC-0300: HDWallets recovery using BIP44
+
+## Abstract
+This ARC proposes the adoption of the BIP44 recovery algorithm for Algorand HDWallets, taking into consideration Algorand's specific use cases such as re-keyed accounts and closed accounts. The BIP44 recovery algorithm provides a standardized approach for deterministic key generation and account recovery, which can enhance the security and usability of Algorand wallets.
+
+## Motivation
+By adopting the BIP44 recovery algorithm, Algorand can provide its users with a reliable and standardized mechanism for account recovery. The BIP44 recovery algorithm is widely used in the cryptocurrency space and has been proven to be effective in ensuring the security and recoverability of user accounts. By adopting this algorithm, Algorand can align itself with industry best practices and provide its users with a seamless and secure account recovery experience.
+
+## Proposal
+The proposal suggests implementing the BIP44 recovery algorithm for Algorand wallets, with the following considerations:
+
+### Re-keyed Accounts
+Algorand supports re-keying of accounts, allowing users to change the signing key associated with their accounts. The BIP44 recovery algorithm should be extended to handle re-keyed accounts, ensuring that the recovered accounts reflect the latest signing key.
+
+### Closed Accounts
+Algorand also supports closing accounts, which means that an account can be closed and no longer being present in the ledger state. The algorithm should consider closed accounts that have been used in the past as part of the scanning process for account recovery.
+
+## Specification
+
+Based on [BIP44 Account Discovery](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#user-content-Account_discovery), we define Algorand's BIP44 path as follows:
+
+```
+m / purpose' / coin_type' / account' / change / address_index
+```
+
+Where:
+- `purpose` is a constant set to `44'` (or `0x8000002C`).
+- `coin_type` is a constant set to `283'` (or `0x8000011B`).
+- `account` is the account index (0-based).
+- `change` is a constant set to `0` for external addresses and `1` for internal addresses.
+- `address_index` is the address index (0-based).
+
+and where `'` denotes a hardened derivation. 
+
+### Definitions
+
+- `address`: Is the 58 characters, base32 encoded representation of an public key.
+- `account`: Is the index of the account in the BIP44 path.
+
+### Account Discovery
+
+1. Starting with `account` index 0
+2. Scan until no **activity** is found for `gap_limit` consecutive addresses.
+    -  If no **activity** is found on the first scan of the account, stop scanning.
+    - If addresses have been found to been re-keyed, the wallet should maintain a map of the found address and the corresponding new address / public which is not expected to sign transactions. 
```

Comment:
We should clarify addresses that have been rekeyed **to** an account address vs account addresses that have been rekeyed. There are three main scenarios. 

1. An external address is rekeyed to an account address within the search. For example, `account[19]` address has no activity, but there IS an Algorand address rekeyed to `account[19]`. Do we expect the wallet to find this? If so, should that count as activity for `account[19]`? 

2. An account address is rekeyed to another account address. In this scenario the wallet should just keep a mapping internally and note the auth addr to the user.

3. An account address is rekeyed to an external address. In this case the wallet should inform the user that the address is rekeyed, but still count it as activity.

Reply:
I'm with Joe here, this second bullet point needs some clarification.

Small nitpick: the following bullet point talks only about the first scan, not the 2nd, 3rd, etc.
```
If no **activity** is found on the first scan of the account, stop scanning.
```

Reply:
Great points @joe-p , will add these cases. 

So for 1) It would consider that an address has **activity**

2) :+1: 

3) :+1:

Reply:
> I'm with Joe here, this second bullet point needs some clarification.
> 
> Small nitpick: the following bullet point talks only about the first scan, not the 2nd, 3rd, etc.
> 
> ```
> If no **activity** is found on the first scan of the account, stop scanning.
> ```

I don't follow @k13n , there's only 1 overall scan. You "scan" until the first 20 addresses (gap_limit) of an account (bip44 account, no Algorand "account") have no activity. Otherwise you keep moving to the next account